src/Symfony/Component/Routing/Router.php

```diff
@@ -162,6 +162,19 @@ public function setOptions(array $options)
         if ($invalid) {
             throw new \InvalidArgumentException(sprintf('The Router does not support the following options: "%s".', implode('", "', $invalid)));
         }
+
+        // This prevent a BC break introduced in 4.3.0
+        // Fallback to PhpGeneratorDumper if the UrlGenerator and UrlGeneratorDumper are not consistent with each other
+        if (is_a($this->options['generator_class'], CompiledUrlGenerator::class, true) !==
+            is_a($this->options['generator_dumper_class'], CompiledUrlGeneratorDumper::class, true)) {
+            $this->options['generator_dumper_class'] = 'Symfony\\Component\\Routing\\Generator\\Dumper\\PhpGeneratorDumper';
```

Comment:
I'm not super comfortable with this: `PhpGeneratorDumper` is deprecated, so that this line of code will have to change in v5. But to what? Let's say we remove it: this means we broke BC without any deprecation notice in v4. To the minimum, this might miss one.
Maybe the added logic is in the wrong method: shouldn't it be in `getGenerator()` instead?
Note also that this overrides any values set by the user. I'd suggest replacing the `is_a` by a direct string comparison instead (that still wouldn't prevent detecting when the option is explicitly set vs loaded from the default)

Reply:
It must be removed when the `PhpGeneratorDumper` will be removed. You're right it misses a notice.

It does not remove all the values set by the user, only the "invalid" ones.
The only cases where values are overriden is when the user use a compiled generator|matcher with a "non-compiled" generator|matcher dumper (inverse is also true)

I think it should be more in `getGeneratorDumperInstance` or `getMatcherDumperInstance`, than `getGenerator` and `getMatcher`. But the changed values won't be visible if you call `getOption`, your thoughts?

I don't think string comparision is right here, as the given classes will in most cases be children of the generator|matcher classes.